### PR TITLE
K fold transform

### PR DESCRIPTION
**Pull request by JenZhao**
I made this to cross validate the baseline_model and the trained model would be different for each fold. We can also change it to validate the trained model with different fold.

### CODE REVIEW

File: hidi/matrix.py [hunk 1]
```diff
@@ -226,6 +227,82 @@ def transform(self, M,  **kwargs):
...
+            if not os.path.exists(newfile_path):
+                os.makedirs(newfile_path)
+            for file in files:
+                if file.startswith('event'):
+                    move(os.path.join(self.log_dir, file),
+                         newfile_path)
+
+            n_fold += 1
+
+        return None, kwargs
```
Comment (kahnjw) [inline hidi/matrix.py]: Does it make sense for this to return the trained model, similar to our non-kfold keras transform?

https://github.com/VEVO/hidi/blob/e266f945f0868fffeeb157da0abf232a16444b9d/hidi/matrix.py#L226

Comment (JenZhao) [inline hidi/matrix.py]: yea. will make the changes

File: hidi/matrix.py [hunk 2]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
         return self.keras_model, kwargs
 
 
+class KerasKfoldTransform(Transform):
+    """
+    Generalized transform for Keras algorithm with k fold cross validation
+    evaluation
+    """
+    def __init__(self, keras_model, validation_matrix, tensorboard_log_dir,
```
Comment (kahnjw) [inline hidi/matrix.py]: Let's remove this function altogether and leave it to the user to configure everything for Keras callbacks. If we start adding built in functionality for it, transforms will get more and more complicated as we make more and more complex keras models.

File: hidi/matrix.py [hunk 3]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
...
+                 classification=False, **keras_kwargs):
+        self.keras_model = keras_model
+        self.keras_kwargs = keras_kwargs
+        self.validation_matrix = validation_matrix
+
+        self.kfold_n_splits = kfold_n_splits
+        self.kfold_seed = kfold_seed
+        self.kfold_shuffle = kfold_shuffle
+
+        self.log_dir = tensorboard_log_dir
```
Comment (kahnjw) [inline hidi/matrix.py]: This can go if we do above.

File: hidi/matrix.py [hunk 4]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
...
+
+            files = os.listdir(self.log_dir)
+            newfile_path = os.path.join(self.log_dir,
+                                        '{0}_fold'.format(n_fold))
+            if not os.path.exists(newfile_path):
+                os.makedirs(newfile_path)
+            for file in files:
+                if file.startswith('event'):
+                    move(os.path.join(self.log_dir, file),
+                         newfile_path)
```
Comment (kahnjw) [inline hidi/matrix.py]: :fire: it

File: hidi/matrix.py [hunk 5]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
...
+        n_fold = 1
+        X = embedding[:, :columns]
+        Y = embedding[:, columns:]
+        for train_index, test_index in kfold.split(X, Y):
+            self.keras_model.fit(
+                X[train_index], Y[train_index],
+                validation_data=[X[test_index], Y[train_index]],
+                **self.keras_kwargs)
+
+            files = os.listdir(self.log_dir)
```
Comment (kahnjw) [inline hidi/matrix.py]: :fire:

File: hidi/matrix.py [hunk 6]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
         return self.keras_model, kwargs
```
Comment (kahnjw) [inline hidi/matrix.py]: `__init__` functions never need return statements, in fact, I think python will just ignore the return (but will evaluate the code in the statement). This like can :fire:

Comment (kahnjw) [inline hidi/matrix.py]: Ah got mixed up, disregard

File: hidi/matrix.py [hunk 7]
```diff
@@ -226,6 +227,74 @@ def transform(self, M,  **kwargs):
...
+
+        if 'item_id' in validation_matrix.columns:
+            self.validation_matrix.set_index('item_id', inplace=True)
+
+    def transform(self, M,  **kwargs):
+        """
+        Takes a Takes a dataframe that has :code:`item_id` index, other
+        'features' columns for prediction, and applies a Keras sequential
+        model to it.
+
```
Comment (kahnjw) [inline hidi/matrix.py]: Add parameter and return value documentation similar to [this code](https://github.com/VEVO/hidi/blob/e266f945f0868fffeeb157da0abf232a16444b9d/hidi/matrix.py#L132-L142).

File: hidi/matrix.py [hunk 8]
```diff
@@ -8,7 +8,9 @@
 from pyvalid import accepts
 from scipy.sparse import csr_matrix
 from sklearn.decomposition import TruncatedSVD
-from sklearn.model_selection import train_test_split
+from sklearn.model_selection import train_test_split, KFold, StratifiedKFold
+from shutil import move
+import os
```
Comment (kahnjw) [inline hidi/matrix.py]: Lets A-Z sort `import` and `from` statements. This should go on line 2.

**Comment by kahnjw** [comment]
The `matrix.py` module is getting pretty big. Starting to think about how we could break it down more, let me know if you have ideas.

File: tests/test_keras_kfoldevaluation_transform.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    def test_baselinemodel_fit_is_called_ten_times(self):
+        t = KerasKfoldTransform(Mock(), self.validation)
+        model, _ = t.transform(self.M)
+        self.assertEqual(10, model.fit.call_count)
+
+    def test_baselinemodel_fit_call_args_shape(self):
+        t = KerasKfoldTransform(Mock(), self.validation)
+        model, _ = t.transform(self.M)
+        x_train, y_train = model.fit.call_args[0]
+        self.assertEqual(x_train.shape, (9000, 6))
```
Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: Will this work?

```python
self.assertEqual(self.M, x_train)
self.assertEqual(self.validation, y_train)
```

Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: Oh, wait no. K-fold, duh

File: tests/test_keras_kfoldevaluation_transform.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        self.M = pd.DataFrame(np.random.rand(10000, 6), index=self.items)
+        self.M.index.name = 'items'
+        self.validation = pd.DataFrame(np.random.rand(10000, 2),
+                                       index=self.items)
+        self.validation.index.name = 'items'
+
+    def test_baselinemodel_fit_is_called_ten_times(self):
+        t = KerasKfoldTransform(Mock(), self.validation)
+        model, _ = t.transform(self.M)
+        self.assertEqual(10, model.fit.call_count)
```
Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: I love how easy dependency injection makes testing :)

Comment (JenZhao) [inline tests/test_keras_kfoldevaluation_transform.py]: should I add more tests here?

Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: Nah. Just add the test:

```python
def test_keras_model_is_returned(self):
    mock = Mock()
    t = KerasKfoldTransform(Mock(), self.validation)
    model, _ = t.transform(self.M)
    self.assertEqual(mock, model)
```

Generally, I try to only have one assert per test, but sometimes I'm lazy.

File: tests/test_keras_kfoldevaluation_transform.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+    def test_baselinemodel_fit_call_args_shape(self):
```
Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: Generally, test names should be assertions, like the one you're made above. So `test_kfold_transform_calls_fit_with_shape_9000x6`

Comment (kahnjw) [inline tests/test_keras_kfoldevaluation_transform.py]: I dont know if I like putting numbers in function names, but there should be a way to make a similar assertion without saying "correct shape" or something redundant like that.

**Comment by kahnjw** [comment]
:+1: I'll cut a release and add a CHANGELOG on master.

**Comment by JenZhao** [comment]
I think we should make a new file just for evaluation related transform, and keep `matrix.py` clean with only matrix factorization related transform

**Comment by kahnjw** [comment]
Cool. Should we land this pull and do that next? I think keeping releases/prs incremental and small is probably a good idea.

**Comment by JenZhao** [comment]
yea sounds great.

**Pull request merged**